src/wallet/api.rs

```diff
@@ -1572,4 +1628,79 @@ impl Wallet {
     pub fn send_tx(&self, tx: &Transaction) -> Result<Txid, WalletError> {
         Ok(self.rpc.send_raw_transaction(tx)?)
     }
+
+    pub fn sweep_incoming_swapcoins(&mut self, feerate: f64) -> Result<Vec<Txid>, WalletError> {
+        let mut swept_txids = Vec::new();
+
+        let completed_swapcoins: Vec<_> = self
+            .store
+            .incoming_swapcoins
+            .iter()
+            .filter_map(|(redeemscript, swapcoin)| {
+                if swapcoin.other_privkey.is_some() {
+                    Some((redeemscript.clone(), swapcoin.clone()))
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        if completed_swapcoins.is_empty() {
+            log::info!("No completed incoming swap coins to sweep");
+            return Ok(swept_txids);
+        }
+
+        log::info!(
+            "Sweeping {} completed incoming swap coins",
+            completed_swapcoins.len()
+        );
+
+        self.sync_no_fail();
+
+        for (multisig_redeemscript, _) in completed_swapcoins {
+            let utxo_info = self
+                .list_incoming_swap_coin_utxo_spend_info()?
+                .into_iter()
+                .find(|(_, spend_info)| {
+                    matches!(spend_info, UTXOSpendInfo::IncomingSwapCoin {
+                        multisig_redeemscript: rs
+                    } if rs == &multisig_redeemscript)
+                });
+
+            if let Some((utxo, spend_info)) = utxo_info {
+                let internal_address = self.get_next_internal_addresses(1)?[0].clone();
+                log::info!(
+                    "Sweeping incoming swap coin {} to internal address {}",
+                    utxo.txid,
+                    internal_address
+                );
+
+                let sweep_tx = self.spend_coins(
+                    &vec![(utxo.clone(), spend_info)],
+                    Destination::Sweep(internal_address.clone()),
+                    feerate,
+                )?;
+                let txid = self.send_tx(&sweep_tx)?;
+
+                self.remove_incoming_swapcoin(&multisig_redeemscript)?;
+                log::info!("Successfully removed incoming swaps coins");
+
+                if self.rpc.get_raw_transaction_info(&txid, None).is_ok() {
+                    swept_txids.push(txid);
+                } else {
+                    log::warn!("Sweep transaction not confirmed, txid: {txid}");
+                }
```

Comment:
Theres a potential problem here. We are deleting the data, before checking if the spend is confirmed. If something goes wrong, the data (money) will be lost.

The checking should be done first, then delete the data.

Also this is only checking for mempool inclusion. We need to wait for confirmation here. Then only its safe to remove the data.

You can put this in a loop waiting for the conf value of the result to reach `Some(1)`.

Reply:
A nice thing to do here would be to isolate the `wait_for_tx_confirmation` into a standalone wallet API method, and call that wherever this type of confirmation waiting is needed. Can be done in a follow-up PR.